### PR TITLE
Refactored Controller Mapping Profile & Moved Input system from core -> sdk

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
- Added Controller Data Providers Profile to register input data sources.
- Controller mapping profiles have been overhauled to be simpler and easier to extend and add new controllers for.
image
- Example of Windows Device Controller Mapping Profile
image
- Updated configuration guide
- addresses #2905 for Input System and components by adding assembly definitions for platform specific code. Future updates when package manager arrives will install/uninstall platform specific stuff as needed.
- addresses #3053 and #3115 for Input System and components.

Breaking Changes
---
- Users will need to reassign the input system service type if using a custom configuration profile.
- Users will need to delete any custom controller mapping profiles, and recreate them.
- Renamed `MixedRealityInputSystemProfile.ControllerMappingProfile` to `MixedRealityInputSystemProfile.ControllerMappingProfiles`
- Renamed `MixedRealityControllerMappingProfile` to `MixedRealityControllerMappingProfiles`

Changes
---
- Fixes: #3074

TODO
---
- [x] #3170

Blocked by
--- 
- [x] #3173
- [x] #3113
- [x] #3242
- [ ] #3148
- [ ] #3254
- [ ] #3234

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
Let's postpone this for review once 2018.3 is released, as the number of options we "Need" to change will be greatly reduced.

**Comment by StephenHodgson** [comment]
Are you referring to Unity 2018.3 release?

Care to explain a bit more?

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+using Microsoft.MixedReality.Toolkit.Core.Definitions.Devices;
+using Microsoft.MixedReality.Toolkit.Core.Services;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Core.Inspectors.Profiles
+{
+    [CustomEditor(typeof(MixedRealityControllerMappingProfiles))]
+    public class MixedRealityControllerMappingProfilesInspector : MixedRealityBaseConfigurationProfileInspector
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs]: Still needs the ability to Add/Remove controller mapping profiles from the list

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                return;
+            }
+
+            if (!MixedRealityToolkit.Instance.ActiveProfile.IsInputSystemEnabled ||
+                 MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.InputActionsProfile == null)
+            {
+                return;
+            }
+
+            mouseControllerMappingProfile = serializedObject.FindProperty("mouseControllerMappingProfile");
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs]: Need to shorten the displayed names, as they are cut off in the inspector.
Or possibly switch the profile line item to a generic image for the platform maybe?
(I really liked the graphical view, which is now hidden behind these line items)

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs]: I don't understand. We no longer have a list.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs]: Agreed. I wasn't quite sure what to do there for that, as we need to "select" the scriptable object profile to get it to display. I'll see what I can do here.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs]: Erm, it looks like a "list" and we do need one.
Almost like the Visualization screen, we should be able to add / remove mappings from the list to meet the requirements for projects, so they can only enable those controllers they want to support. (and easily add more)

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfilesInspector.cs]: I got rid of the list as it was a lot harder to manage the it and prevent multiple from being registered, but I'll see what I can do.

**Comment by StephenHodgson** [comment]
I'll also need to update the current docs to reflect the changes here.

**Comment by StephenHodgson** [comment]
Found an issue with the generic mappings that don't reflect their settings from the profile when opening the advanced input mapping window.

**Comment by StephenHodgson** [comment]
Fixed

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs [hunk 3]
```diff
@@ -115,21 +115,21 @@ public MixedRealitySpeechCommandsProfile SpeechCommandsProfile
...
 
         [SerializeField]
         [Tooltip("Device profile for wiring up physical inputs to Actions.")]
-        private MixedRealityControllerMappingProfile controllerMappingProfile;
+        private MixedRealityControllerMappingProfiles controllerMappingProfiles;
 
         /// <summary>
         /// Active profile for controller mapping configuration
         /// </summary>
-        public MixedRealityControllerMappingProfile ControllerMappingProfile
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs]: Oh, it seems I renamed the whole class. Need to stub that out and mark it obsolete too.

**Comment by SimonDarksideJ** [comment]
Is this ready for review again yet @StephenHodgson ?

**Comment by SimonDarksideJ** [comment]
Just to be awkward, would it better to have one profile per SDK?  
With WMR we have Immersive and HL, we have one Vive profile but multiple Oculus ones.

**Comment by StephenHodgson** [comment]
This is true. But I think the oculus remote was a completely different device, where as the WMR devices are provided through the same API

**Comment by StephenHodgson** [comment]
Yes, this is ready for another review. (sans the Oculus devices)

**Comment by StephenHodgson** [comment]
@SimonDarksideJ It's now one list.

File: Documentation/MixedRealityConfigurationGuide.md [hunk 4]
```diff
@@ -275,23 +275,53 @@ The configuration screen below allows you to configure any of the controllers cu
...
+
+![](/External/HowTo/MixedRealityToolkitConfigurationProfileScreens/MRTK_OculusTouchConfigScreen.png)
 
 > Clicking on the "Back to Configuration Profile" button will take you back to the Mixed Reality Toolkit Input System Settings screen.
+
+## Advanced Controller Mapping Setup
+
+For advanced users who want to add custom input mappings through Unity's input system, you can use the generic Unity device or generic Open VR device profiles.
+
+![](/External/HowTo/MixedRealityToolkitConfigurationProfileScreens/MRTK_AdvancedControllerMapping.png)
```
Comment (StephenHodgson) [inline Documentation/MixedRealityConfigurationGuide.md]:
```suggestion
![](/External/HowTo/MixedRealityToolkitConfigurationProfileScreens/MRTK_AdvancedControllerMapping.PNG)
```

**Comment by SimonDarksideJ** [comment]
That's brilliant @StephenHodgson and just what I was looking for.
One comment though, can we shorten the names listed ?
image
To just what they are for, e.g. Mouse, Touchscreen, Xbox, Generic OpenVR, etc.

We already know they are "Controller Mapping Profiles", so no need to state it for each item :D

**Comment by StephenHodgson** [comment]
Sure I can remove "controller mapping profile" from the end of the name

**Comment by SimonDarksideJ** [comment]
Sent you a PR doing just that :D

**Comment by StephenHodgson** [comment]
Going to update this PR with the changes discussed for Devices

**Comment by StephenHodgson** [comment]
Blocked by #3148 so I can get the spatial observer out of the controller folder.

**Pull request closed without merging**